sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala

```diff
@@ -86,4 +87,8 @@ class FileScanRDD(
   }
 
   override protected def getPartitions: Array[Partition] = filePartitions.toArray
+
+  override protected def getPreferredLocations(split: Partition): Seq[String] = {
+    split.asInstanceOf[FilePartition].files.flatMap(_.locations).distinct
```

Comment:
Looks we want to pick the location based on the size of a block? In MapOutputTracker we have `getPreferredLocationsForShuffle`, which can be related to us.